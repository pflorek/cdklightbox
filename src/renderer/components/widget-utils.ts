import {
  AnyWidgetState,
  ConstructWidgetState,
  getRef,
  Parameter,
  StaticFunctionWidgetState,
  WidgetState,
  WorkbenchState,
} from '../state';
import { PickedWidget } from './Picker';

function getStackOrOpen(widgetState: WidgetState): WidgetState {
  let currentWidgetState = widgetState;
  const isStackClass = (forType: string) => forType === '@aws-cdk/core.Stack';

  while (!isStackClass(currentWidgetState.forType)) {
    currentWidgetState = currentWidgetState.scope as WidgetState;
  }
  return currentWidgetState;
}

function isStackOrOpen(widgetState: WidgetState) {
  return !!getStackOrOpen(widgetState);
}

export const handleValuePicked = (
  addWidget: (
    newWidget: WidgetState,
    refUpdater?: (state: WorkbenchState, newKey: string) => void
  ) => void,
  updateWidget: (
    updater: (
      widgets:
        | Record<string, WidgetState>
        | Record<string, StaticFunctionWidgetState>
        | Record<string, ConstructWidgetState>
    ) => void
  ) => void,
  removePicker: (picker: string) => void,
  widgetState: WidgetState,
  widgetKey: string,
  parametersProvider: (widgetState: AnyWidgetState) => Record<string, Parameter>
) => {
  return (props: PickedWidget) => {
    if (props.widget) {
      const newWidget = {
        ...props.widget,
        scope: isStackOrOpen(widgetState)
          ? widgetState.scope
          : getStackOrOpen(widgetState).scope,
      };
      addWidget(newWidget, (workbench: WorkbenchState, newKey: string) => {
        parametersProvider(workbench.widgets[widgetKey])[props.property.name] =
          {
            ref: getRef(newKey),
            widgetShapes: {},
          };
      });
    }
    if (props.value !== undefined) {
      updateWidget((widgets) => {
        const currentWidget = widgets[widgetKey];
        parametersProvider(currentWidget)[props.property.name] = {
          value: props.value,
          widgetShapes: {},
        };
      });
    }
    if (props.pickerKey) {
      removePicker(props.pickerKey);
    }
  };
};

export const handleValueZipped = (
  zip,
  unzip,
  constructKey: string,
  parametersProvider: (widgetState: AnyWidgetState) => Record<string, Parameter>
) => {
  return (prop: any, isZipped: boolean) => {
    if (isZipped) {
      zip(
        (state) => parametersProvider(state.widgets[constructKey])[prop.name]
      );
    } else {
      unzip(
        (state) => parametersProvider(state.widgets[constructKey])[prop.name]
      );
    }
  };
};
