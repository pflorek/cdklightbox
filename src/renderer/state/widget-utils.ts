import {
  EnumWidgetState,
  StaticFunctionWidgetState,
  WidgetState,
} from './states';
import { jsii } from '../jsii';

export function isStaticFunction(widgetState: StaticFunctionWidgetState) {
  return !!(widgetState.method && widgetState.parameters);
}

export function isEnum(widgetState: EnumWidgetState) {
  return !!(widgetState.method && widgetState.type === 'enum');
}

export function isConstruct(widgetState: WidgetState) {
  if ((widgetState as StaticFunctionWidgetState).type === 'staticFunction') {
    return false;
  }
  if ((widgetState as EnumWidgetState).type === 'enum') {
    return false;
  }
  const jsiiType = jsii.types[widgetState.forType];

  return !!(
    jsiiType.initializer?.parameters &&
    jsiiType.initializer?.parameters[0].type.fqn === 'constructs.Construct'
  );
}

export function isLooseObject(widgetState: WidgetState) {
  return (
    !isStaticFunction(widgetState as StaticFunctionWidgetState) &&
    !isEnum(widgetState as EnumWidgetState) &&
    !isConstruct(widgetState)
  );
}

export function getRef(key: string) {
  return `WIDGET#${key}`;
}

export enum WidgetTypeShape {
  LOOSE = 'LOOSE',
  STATIC_FUNCTION = 'STATIC_FUNCTION',
  ENUM = 'ENUM',
}

export function getKey(ref: string) {
  return ref.replace('WIDGET#', '');
}

export function isRef(key: string) {
  return key.startsWith('WIDGET#');
}
