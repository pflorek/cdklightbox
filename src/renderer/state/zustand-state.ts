import create from 'zustand';
import { persist, subscribeWithSelector } from 'zustand/middleware';
import { Set, WorkbenchState } from './states';
import { errorState } from './substates/error-state';
import { workingDirectoryState } from './substates/working-directory-state';
import { widgetsState } from './substates/widget-state';
import { statusState } from './substates/status-state';

// eslint-disable-next-line import/prefer-default-export
export const useWorkbenchStore = create<WorkbenchState>(
  // @ts-ignore
  subscribeWithSelector(
    persist(
      (set: Set) => ({
        ...errorState(set),
        ...statusState(set),
        ...workingDirectoryState(set),
        ...widgetsState(set),
        loadState(newState: Partial<WorkbenchState>) {
          set(() => newState);
        },
      }),

      {
        name: 'localStorage',
        getStorage: () => localStorage,
      }
    )
  )
);
