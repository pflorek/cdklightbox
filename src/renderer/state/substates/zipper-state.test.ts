import * as fs from 'fs';
import * as path from 'path';
import {
  ConstructWidgetState,
  LooseObjectWidgetState,
  StaticFunctionWidgetState,
  WorkbenchState,
} from '../states';
import { getRef, WidgetTypeShape } from '../widget-utils';
import { useWorkbenchStore } from '../zustand-state';

describe('zip', () => {
  it('moves references', () => {
    // given
    const constructKey = 'myconstruct';
    const staticFunctionKey = 'myStaticFunction';
    const otherLooseObjectKey = 'otherlooseobject';
    const vpcConstructKey = 'avpc';

    const otherConstructProps = {
      scope: {},
      id: '',
      x: 0,
      y: 0,
      forType: '',
    };
    // @ts-ignore
    const construct: ConstructWidgetState = {
      ...otherConstructProps,
      initializerParameters: {
        something: { ref: getRef(staticFunctionKey), widgetShapes: {} },
      },
    };
    const staticFunction: StaticFunctionWidgetState = {
      id: '',
      x: 0,
      y: 0,
      method: { name: '', parameters: [] },
      forType: '@aws-cdk/aws-ec2.InstanceType',
      type: 'staticFunction',
      parameters: {
        somethingElse: { ref: getRef(otherLooseObjectKey) },
        notARef: { value: 'something else' },
      },
    };
    const otherLooseObject: LooseObjectWidgetState = {
      forType: '@aws-cdk/aws-ec2.InstanceType',
      id: '',
      x: 0,
      y: 0,
      initializerParameters: {
        vpc: {
          ref: getRef(vpcConstructKey),
          widgetShapes: {},
        },
      },
    };
    const vpcConstruct: ConstructWidgetState = {
      // @ts-ignore
      scope: {},
      id: '',
      x: 0,
      y: 0,
      initializerParameters: {},
      forType: '@aws-cdk/aws-ec2.Vpc',
    };

    useWorkbenchStore.setState({
      widgets: {
        [constructKey]: construct,
        [staticFunctionKey]: staticFunction,
        [otherLooseObjectKey]: otherLooseObject,
        [vpcConstructKey]: vpcConstruct,
      },
      pickers: {},
    });

    // when
    useWorkbenchStore
      .getState()
      .zip(
        (state) =>
          (state.widgets[constructKey] as ConstructWidgetState)
            .initializerParameters.something
      );

    // then
    const updatedState: WorkbenchState = useWorkbenchStore.getState();
    expect(updatedState.widgets[staticFunctionKey]).toBeTruthy();
    expect(
      (updatedState.widgets[staticFunctionKey] as StaticFunctionWidgetState)
        .parameters.somethingElse.widgetShapes[otherLooseObjectKey]
    ).toEqual(WidgetTypeShape.LOOSE);
    expect(updatedState.widgets[otherLooseObjectKey]).toBeTruthy();
    expect(
      (updatedState.widgets[constructKey] as ConstructWidgetState)
        .initializerParameters.something.widgetShapes[staticFunctionKey]
    ).toEqual(WidgetTypeShape.STATIC_FUNCTION);
    expect(
      (updatedState.widgets[constructKey] as ConstructWidgetState)
        .initializerParameters.something.widgetShapes[otherLooseObjectKey]
    ).toEqual(WidgetTypeShape.LOOSE);
    expect(updatedState.widgets[vpcConstructKey].zipped).toBeFalsy();
  });

  it('More complete example', () => {
    const workbenchState: WorkbenchState = JSON.parse(
      fs.readFileSync(
        path.join(__dirname, 'test_states', 'workbench.json'),
        'utf-8'
      )
    );
    // @ts-ignore
    const INSTANCE_KEY = Object.entries(workbenchState.widgets).find(
      ([, state]) => state.forType === '@aws-cdk/aws-ec2.Instance'
    )[0];
    useWorkbenchStore.setState(() => workbenchState);
    useWorkbenchStore
      .getState()
      .zip(
        (state) =>
          (state.widgets[INSTANCE_KEY] as ConstructWidgetState)
            .initializerParameters.vpc
      );

    expect(
      (
        useWorkbenchStore.getState().widgets[
          INSTANCE_KEY
        ] as ConstructWidgetState
      ).initializerParameters.vpc.widgetShapes
    ).toMatchSnapshot();
  });
});

describe('unzip', () => {
  it('moves references back', () => {
    // given
    const constructKey = 'myconstruct';
    const staticFunctionKey = 'myStaticFunction';
    const otherLooseObjectKey = 'otherlooseobject';
    const otherLooseObjectProps = {
      initializerParameters: {},
      forType: '',
      id: '',
      method: '',
      type: '',
      x: 0,
      y: 0,
    };
    const otherLooseObject: LooseObjectWidgetState = {
      ...otherLooseObjectProps,
      zipped: 2,
    };
    const otherStaticFunctionProps: Partial<StaticFunctionWidgetState> = {
      x: 0,
      y: 0,
      id: '',
      type: 'staticFunction',
      forType: 'somestaticfunctionfqn',
      method: { name: 'somemethod', parameters: [] },
    };

    // @ts-ignore
    const staticFunction: StaticFunctionWidgetState = {
      parameters: {
        somethingElse: {
          ref: getRef(otherLooseObjectKey),
          widgetShapes: { [otherLooseObjectKey]: WidgetTypeShape.LOOSE },
        },
        notARef: { value: 'something else' },
      },
      ...otherStaticFunctionProps,
      zipped: 1,
    };

    const construct: ConstructWidgetState = {
      // @ts-ignore
      scope: {},
      id: '',
      x: 0,
      y: 0,
      forType: 'someconstructsfqn',
      initializerParameters: {
        something: {
          ref: getRef(staticFunctionKey),
          widgetShapes: {
            [staticFunctionKey]: WidgetTypeShape.STATIC_FUNCTION,
          },
        },
      },
    };

    useWorkbenchStore.setState({
      widgets: {
        [constructKey]: construct,
        [staticFunctionKey]: staticFunction,
        [otherLooseObjectKey]: otherLooseObject,
      },
      pickers: {},
    });

    // when
    useWorkbenchStore
      .getState()
      .unzip(
        (state) =>
          (state.widgets[constructKey] as ConstructWidgetState)
            .initializerParameters.something
      );

    // then
    const updatedState: WorkbenchState = useWorkbenchStore.getState();
    expect(
      (updatedState.widgets[constructKey] as ConstructWidgetState)
        .initializerParameters.something.widgetShapes[staticFunctionKey]
    ).toBeUndefined();
    expect(updatedState.widgets[staticFunctionKey]).toEqual({
      parameters: {
        somethingElse: {
          ref: getRef(otherLooseObjectKey),
          widgetShapes: {},
        },
        notARef: { value: 'something else' },
      },
      ...otherStaticFunctionProps,
    });
    expect(updatedState.widgets[otherLooseObjectKey]).toEqual({
      ...otherLooseObjectProps,
    });
  });
});
