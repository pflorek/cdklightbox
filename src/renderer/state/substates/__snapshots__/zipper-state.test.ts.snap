// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`zip More complete example 1`] = `
Object {
  "9viOTBZmqJVJMlPxKAqGn": "LOOSE",
  "naP4UUEniUtn4qI7I-22D": "STATIC_FUNCTION",
}
`;
